conbench/entities/distribution.py

```diff
@@ -108,109 +115,122 @@ def get_distribution(benchmark_result, limit):
-                set_=values,
-            )
-        )
-        conn.commit()
-
-
-def get_closest_parent(run):
-    commit = run.commit
-    if commit.timestamp is None:
-        return None
-
-    hardware_entities = Hardware.all(hash=run.hardware.hash)
-    hardware_ids = set([m.id for m in hardware_entities])
-
-    # TODO: what about matching contexts
-    result = (
-        Session.query(
-            Run.id,
-            Commit.id,
-        )
-        .join(Commit, Commit.id == Run.commit_id)
-        .join(Hardware, Hardware.id == Run.hardware_id)
-        .filter(
-            Run.reason == "commit",
-            Hardware.id.in_(hardware_ids),
-            Commit.timestamp.isnot(None),
-            Commit.timestamp < commit.timestamp,
-            Commit.repository == commit.repository,
+    values["limit"] = commit_limit
+
+    Session.execute(
+        insert(Distribution.__table__)
+        .values(values)
+        .on_conflict_do_update(
+            index_elements=["case_id", "context_id", "commit_id", "hardware_hash"],
+            set_=values,
         )
-        .order_by(Commit.timestamp.desc())
-        .first()
     )
-
-    parent = None
-    if result:
-        commit_id = result[1]
-        parent = Commit.get(commit_id)
-
-    return parent
-
-
-def set_z_scores(benchmark_results):
-    if not benchmark_results:
-        return
+    Session.commit()
+    return values
+
+
+def set_z_scores(benchmark_results: List["BenchmarkResult"]):
+    """Populate the "z_score" attribute of each given BenchmarkResult, comparing it to
+    the distribution of BenchmarkResults that share the same hardware/case/context,
+    ending with the most recent BenchmarkResult in the given BenchmarkResult's commit
+    ancestry on the default branch.
+
+    We choose a commit from the default branch as the end commit, because we probably
+    don't want "bad" commits early on in a PR to set artificially low z-scores later in
+    the PR.
+
+    For the start commit of the distribution, uses whatever commit_limit was used when
+    writing to the Distribution table.
+
+    Should not raise anything. If we can't find a z-score for some reason, the z_score
+    attribute will be None on that BenchmarkResult.
+    """
+    # Cache Distribution query results by commit/hardware, in case multiple input
+    # benchmark_results (e.g. from the same Run) share those attributes
+    cached_distributions = {}
+    # Also cache respective baseline runs by each Run's ID. Note that two
+    # benchmark_results from the same Run might not have the exact same baseline run,
+    # because one baseline run might be missing the case/context of the other
+    # benchmark_result. In those cases, we might fail to populate the z-score of the
+    # second benchmark result. But that's too rare and small-impact to justify not
+    # caching for performance reasons.
+    cached_baseline_runs = {}
```

Comment:
Yes, this is correct. One clarification to your "IIUC" parenthetical: a person or the webapp could hit `/api/benchmarks` for the last 500 benchmarks regardless of run, or `/api/benchmarks?name=X` which also disregards run, or `/api/benchmarks?batch_id=X` which could theoretically return benchmarks from multiple runs (since there's currently no technical restriction that one batch belongs to one run). That's why we're handling potentially-multiple runs in this function.

I agree that loosening this would be nice, so that each benchmark finds its own baseline. That's actually how I had originally written this function before I changed it in my latest commit. The problem was it was taking a really long time to `get_baseline_run()` for every single benchmark. There's probably a smart middle ground where we do less-aggressive caching.

Reply:
Made #567